### PR TITLE
Add support for launching activity aliases

### PR DESCRIPTION
**Pull request by sschuberth**
Unfortunately, aapt does not return a "launchable-activity" property if
the launchable activity is an activity alias. Work around that by manually
parsing aapt's pseudo-XML output.

EDIT: Added two more commits, one to fix issue #74, and one to allow app downgrades when installing.

### CODE REVIEW

**Comment by hal9002** [comment]
Can one of the admins review this PR?

**Comment by stefanhoth** [comment]
Retest this
On Jun 26, 2015 2:24 PM, "hal9002" <notifications@github.com> wrote:

> Can one of the admins review this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/novoda/gradle-android-command-plugin/pull/79#issuecomment-115659695>
> .
>

-- 
Disclaimer: The information in this e-mail and any attachments is the 
property of Novoda Ltd and is confidential and may be legally privileged. 
This e-mail is intended solely for the person or organisation to which it 
is addressed. Any disclosure, copying or other use of the information by 
any person or organisation who is not the intended recipient is strictly 
prohibited and may be unlawful. If you have received this e-mail in error, 
please inform the sender immediately and delete/destroy this e-mail and any 
copies of it. Novoda Ltd has taken reasonable precautions to minimise the 
risk of any software viruses which may damage your systems, but we advise 
that you take the necessary steps to ensure that no virus contamination is 
suffered. Novoda Ltd does not accept any liability for any loss or damage 
caused by the transmission of any virus.

Novoda Ltd, Company No: 347444, Registered in Scotland Registered 
Office: C/O Alexander Sloan, 38 Cadogan Street, Glasgow, G2 7HF, Scotland.  VAT 
Registration Number  GB 984 2525 93

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/gradle-android-command-plugin-prb/1/
Test FAILed.

**Comment by sschuberth** [comment]
To me this seem to be a CI issue as the Jenkins log says:

    ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.

**Comment by stefanhoth** [comment]
Hmmm, it may be that the PRB can't handle remote branches. That would be weird.

**Comment by hal9002** [comment]
Can one of the admins review this PR?

**Comment by hal9002** [comment]
Can one of the admins review this PR?

**Comment by stefanhoth** [comment]
@ouchadam did you try PRB on remote branches yet?

**Comment by stefanhoth** [comment]
Ok to test.

**Comment by ouchadam** [comment]
It should just work... (After a white listed person triggers the build)

**Comment by stefanhoth** [comment]
@ouchadam it can't find the revision. I thought it may be because the commit does not exist in our repo.

**Comment by sschuberth** [comment]
As PRs live in a special refs namespace, you probably need to configure the Jenkins Git plugin to fetch these?

**Comment by ouchadam** [comment]
Could be the job looking at the master branch and not a refspec
On 26 Jun 2015 13:36, "Stefan Hoth" <notifications@github.com> wrote:

> @ouchadam <https://github.com/ouchadam> it can't find the revision. I
> thought it may be because the commit does not exist in our repo.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/novoda/gradle-android-command-plugin/pull/79#issuecomment-115663436>
> .
>

-- 
Disclaimer: The information in this e-mail and any attachments is the 
property of Novoda Ltd and is confidential and may be legally privileged. 
This e-mail is intended solely for the person or organisation to which it 
is addressed. Any disclosure, copying or other use of the information by 
any person or organisation who is not the intended recipient is strictly 
prohibited and may be unlawful. If you have received this e-mail in error, 
please inform the sender immediately and delete/destroy this e-mail and any 
copies of it. Novoda Ltd has taken reasonable precautions to minimise the 
risk of any software viruses which may damage your systems, but we advise 
that you take the necessary steps to ensure that no virus contamination is 
suffered. Novoda Ltd does not accept any liability for any loss or damage 
caused by the transmission of any virus.

Novoda Ltd, Company No: 347444, Registered in Scotland Registered 
Office: C/O Alexander Sloan, 38 Cadogan Street, Glasgow, G2 7HF, Scotland.  VAT 
Registration Number  GB 984 2525 93

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Install.groovy [hunk 1]
```diff
@@ -6,6 +6,6 @@ class Install extends AdbTask {
 
     @TaskAction
     void exec() {
-        assertDeviceAndRunCommand(['install', '-r', apkPath])
+        assertDeviceAndRunCommand(['install', '-r', '-d', apkPath])
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Install.groovy]: Why the `-d`?

Comment (sschuberth) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Install.groovy]: See the commit message :-) My commit message usually contain valuable information, so you should not rely on reading just the combined diff of all commits.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy [hunk 2]
```diff
@@ -70,7 +110,7 @@ public class AdbTask extends org.gradle.api.DefaultTask {
             it.startsWith("$propertyKey:")
         }
         if (output == null) {
-            throw new IllegalStateException("Could not read property '$propertyKey' of $apkPath")
+            throw new IllegalArgumentException("Could not read property '$propertyKey' of $apkPath")
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: Why changed?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/VariantConfigurator.groovy [hunk 3]
```diff
@@ -33,7 +33,7 @@ class VariantConfigurator {
             task.group = AndroidCommandPlugin.TASK_GROUP + " for variant " + variationName;
         }
 
-        task.apkPath = "${-> variant.outputs[0].packageApplication.outputFile}"
+        task.apkPath = "${-> variant.outputs[0].outputFile}"
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/VariantConfigurator.groovy]: Has that changed?

Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/VariantConfigurator.groovy]: Ok, read your commit message now. ;)

Comment (sschuberth) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/VariantConfigurator.groovy]: For reference, no, I believe it has not changed in the Android Gradle plugin. The gradle-android-command-plugin just always used the wrong API ;-)

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy (hunk 2 above)
Comment (sschuberth) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: Two reasons: First, I need the caller to be able to distinguish the `output == null` case from the other IllegalStateException thrown in line 107 (of the new code). Second, IMO an IllegalStateException is not semantically correct in this case. If `output == null` it does not mean the plugin is in a wrong state, but that there was no match for the `propertyKey` argument, hence an IllegalArgumentException seems more appropriate to me.

Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: Ah, okay, I see. 
I wonder whether we should better just return `null` instead. It does not seem like an exceptional state after all. 
Could then just use `?.` in line 31.
What do you think?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Install.groovy [hunk 4]
```diff
@@ -6,6 +6,6 @@ class Install extends AdbTask {
 
     @TaskAction
     void exec() {
-        assertDeviceAndRunCommand(['install', '-r', apkPath])
+        assertDeviceAndRunCommand(['install', '-rd', apkPath])
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Install.groovy]: :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy (hunk 2 above)
Comment (sschuberth) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: I had been thinking about that, too, but was a bit undecided. Also line 23 (`getPackageName()`) would require fixing. I'll push a version that returns `null` anyway, and then you can choose :-)

**Comment by sschuberth** [comment]
@devisnik Do you like this `null`-check version better than the one that keeps throwing the exception?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy [hunk 5]
```diff
@@ -20,16 +20,63 @@ public class AdbTask extends org.gradle.api.DefaultTask {
...
-        if (matcher) {
-            matcher[0][1]
+        def output = readApkProperty('package')
+        if (output) {
+            def matcher = output.readLines()[0] =~ /name='(.*?)'/
+            if (matcher) {
+                matcher[0][1]
+            }
+        } else {
+            throw new IllegalArgumentException("Could not read property '$propertyKey' of $apkPath")
```
Comment (friedger) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: Should be "Could not read property 'package' of $apkPath

**Comment by devisnik** [comment]
`null` Version looks better. :+1: 
Good to merge now, imho.
I'll probably base some refactoring on this later on.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy (hunk 5 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: Ah, yes.

Comment (sschuberth) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AdbTask.groovy]: Thanks, fixed.

**Comment by devisnik** [comment]
Great addition, thanks for contributing @sschuberth .

**Comment by sschuberth** [comment]
You're welcome, thanks for the prompt reviews. Can we soon expect a new release with the added activity-alias support?

**Comment by friedger** [comment]
@sschuberth On its way .. https://github.com/novoda/gradle-android-command-plugin/pull/82

**Pull request merged**